src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,682 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A type interface for a range of bits.
+#[allow(clippy::exhaustive_structs)]
+#[derive(Debug, Clone)]
+pub struct BitRange<'a, T, const START: u8, const END: u8>(pub &'a T);
+
+/// A type interface for a range of bits.
+#[allow(clippy::exhaustive_structs)]
+#[derive(Debug)]
+pub struct BitRangeMut<'a, T, const START: u8, const END: u8>(pub &'a mut T);
+
+// std::fmt::Display
+macro_rules! bit_range_display {
+    ($x:ty) => {
+        impl<const START: u8, const END: u8> std::fmt::Display for BitRange<'_, $x, START, END> {
+            #[doc = concat!("
+                ```
+                use bit_fields::BitRange;
+                let x = 18", stringify!($x), ";
+                assert_eq!(BitRange::<_,0,4>(&x).to_string(),2", stringify!($x),".to_string());
+                assert_eq!(BitRange::<_,4,8>(&x).to_string(),1", stringify!($x),".to_string());
+                ```
+            ")]
+            #[inline]
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}", <$x>::from(self))
+            }
+        }
+    };
+}
+
+bit_range_display!(u128);
+bit_range_display!(u64);
+bit_range_display!(u32);
+bit_range_display!(u16);
+bit_range_display!(u8);
+
+macro_rules! bit_range_mut_display {
+    ($x:ty) => {
+        impl<const START: u8, const END: u8> std::fmt::Display for BitRangeMut<'_, $x, START, END> {
+            #[doc = concat!("
+                ```
+                use bit_fields::BitRangeMut;
+                let mut x = 18", stringify!($x), ";
+                assert_eq!(BitRangeMut::<_,0,4>(&mut x).to_string(),2", stringify!($x),".to_string());
+                assert_eq!(BitRangeMut::<_,4,8>(&mut x).to_string(),1", stringify!($x),".to_string());
+                ```
+            ")]
+            #[inline]
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}", <$x>::from(self))
+            }
+        }
+    };
+}
+
+bit_range_mut_display!(u128);
+bit_range_mut_display!(u64);
+bit_range_mut_display!(u32);
+bit_range_mut_display!(u16);
+bit_range_mut_display!(u8);
+
+// impl BitRange<..>
+macro_rules! bit_range {
+    ($x:ty,$mask:ident,$max:ident) => {
+        impl<const START: u8, const END: u8> BitRange<'_, $x, START, END> {
+            pub const MASK: $x = $mask::<START, END>();
+            /// The maximum value this range can store
+            pub const MAX: $x = $max::<START, END>();
```

Comment:
You can get rid of `shift!`, `mask_fn!` and `max_fn!` like this

```suggestion
const fn ones(how_many: u8, significant: u32) -> u128 {
    if how_many as u32 >= significant {
        u128::MAX
    } else {
        (1 << how_many as u32) - 1
    }
}

// impl BitRange<..>
macro_rules! bit_range {
    ($x:ty,$mask:ident,$max:ident) => {
        impl<const START: u8, const END: u8> BitRange<'_, $x, START, END> {
            pub const MASK: $x = (ones(END, <$x>::BITS) - ones(START, <$x>::BITS)) as $x;
            /// The maximum value this range can store
            pub const MAX: $x = ones(END - START, <$x>::BITS) as $x;
```

Reply:
I think this adds complexity without visible benefit. A few simple const functions are more readable in my opinion.